### PR TITLE
Custom sliders

### PR DESCRIPTION
**Pull request by ddevault**
Depends on #214, https://github.com/ppy/osu-resources/pull/9, and https://github.com/ppy/osu-framework/pull/384.

Still could use some tweaking but it's a good start, I think.

### CODE REVIEW

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 1]
```diff
@@ -1,53 +1,100 @@
...
+                text = new SpriteText { Alpha = 0 },
+                slider = new OsuSliderBar<T>
+                {
+                    Margin = new MarginPadding { Top = 5 },
+                    RelativeSizeAxes = Axes.X,
+                }
+            };
+        }
+
+        private class OsuSliderBar<U> : SliderBar<U> where U : struct,
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: i realy dislike this formatting. why not move the whole `where U...` to a new line so it's obvious that everything following it is related to U?

**Comment by peppy** [comment]
I'm going to have to assume that you don't test anything you code at this rate. Brings back memories of someone else who we had to stop working on the codebase due to the path of destruction they left behind them.

image



image



border is too thin compared to line
image

custom design relies on fucking with base class too much. best off making it not rely on the base framework design through other means (see checkbox implementation maybe).

**Comment by peppy** [comment]
if you can't figure how to do this right, there's a completely working and simple implementation in the lazer transitional codebase i wrote. it should be quite easily portable to here

**Comment by peppy** [comment]
- sound effect doesn't play

**Comment by peppy** [comment]
- left and right arrow key both move right

**Comment by SirCmpwn** [comment]
>sound effect doesn't play

osu-resources up to date? works for me

>left and right arrow key both move right

Almost certainly unrelated to my changes, but I'll look into it anyway

>custom design relies on fucking with base class too much. best off making it not rely on the base framework design through other means (see checkbox implementation maybe).

Fair enough

>if you can't figure how to do this right, there's a completely working and simple implementation in the lazer transitional codebase i wrote. it should be quite easily portable to here

I kind of doubt it'll port that easily but I'll have a look

- - -

Regarding the other issues, I knew they were there. Like I said, this is just a start and more work is required. I guess I'll look into fixing them now, though.

**Comment by Tom94** [comment]
This is wrong: https://streamable.com/28ar

Also, left and right arrow keys seem to control the wrong sliderbars disregarding which one is hovered: https://streamable.com/np40

**Comment by Tom94** [comment]
Also, I think the sliderbar should perfectly follow the mouse movement, and not lag behind via a delayed transformation. The smoothed movement should only be used with keyboard input.

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 2]
```diff
@@ -1,53 +1,155 @@
-﻿using System;
+﻿using System;
```
Comment (Tom94) [inline osu.Game/Overlays/Options/SliderOption.cs]: Missing license header.

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 3]
```diff
@@ -1,53 +1,155 @@
...
+            {
+                playSample();
+                return base.OnDrag(state);
+            }
+
+            protected override void UpdateValue(float value)
+            {
+                nub.MoveToX(DrawWidth * value, 300, EasingTypes.OutQuint);
+                leftBox.ScaleTo(new Vector2(
+                        MathHelper.Clamp(DrawWidth * value - nub.Width / 2 + 2, 0, DrawWidth),
```
Comment (Tom94) [inline osu.Game/Overlays/Options/SliderOption.cs]: This clamp is the reason for the first issue I pointed out earlier. I recommend you only apply a transformation to the nub, and inside an overridden `Update` you update the scales of the right and left boxes based on the nub position.

**Comment by SirCmpwn** [comment]
>Also, left and right arrow keys seem to control the wrong sliderbars disregarding which one is hovered: https://streamable.com/np40

We should talk about this. What's the best solution? I could stash the mouse state and check if it's over our component, but we should probably think about improving focus handling at a broader level (and in seperate PRs).

Other feedback has been addressed.

**Comment by peppy** [comment]
merge master

**Comment by peppy** [comment]
can't scroll properly. check textbox logic if you're unsure how to fix this.

**Comment by peppy** [comment]
padding wrong

image

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 4]
```diff
@@ -1,53 +1,163 @@
...
-                    SelectionColor = new Color4(255, 102, 170, 255),
-                }
-            };
-        }
-    }
-}
\ No newline at end of file
+            protected override void UpdateValue(float value)
+            {
+                nub.Position = new Vector2(DrawWidth * value, nub.Position.Y);
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: `nub.MoveToX(DrawWidth * value);`

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 5]
```diff
@@ -1,53 +1,163 @@
...
+            protected override bool OnDrag(InputState state)
+            {
+                playSample();
+                return base.OnDrag(state);
+            }
+            
+            protected override void Update()
+            {
+                base.Update();
+                leftBox.Scale = new Vector2(MathHelper.Clamp(
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: use DrawWidth instead of DrawSize.X?

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 6]
```diff
@@ -1,53 +1,163 @@
...
+            protected override bool OnClick(InputState state)
+            {
+                playSample();
+                return base.OnClick(state);
+            }
+            
+            protected override bool OnDrag(InputState state)
+            {
+                playSample();
+                return base.OnDrag(state);
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: if we're handling drag here then we `return true`

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 7]
```diff
@@ -1,53 +1,163 @@
...
+                    sample.Frequency.Value = 1 + NormalizedValue * 0.2f;
+                    sample.Play();
+                }
+            }
+            
+            protected override bool OnKeyDown(InputState state, KeyDownEventArgs args)
+            {
+                if (args.Key == Key.Left || args.Key == Key.Right)
+                    playSample();
+                return base.OnKeyDown(state, args);
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: return true

File: osu.Game/Overlays/Options/SliderOption.cs (hunk 6 above)
Comment (SirCmpwn) [inline osu.Game/Overlays/Options/SliderOption.cs]: base.OnDrag does so

File: osu.Game/Overlays/Options/SliderOption.cs (hunk 7 above)
Comment (SirCmpwn) [inline osu.Game/Overlays/Options/SliderOption.cs]: ditto

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 8]
```diff
@@ -1,53 +1,163 @@
...
+
+            [BackgroundDependencyLoader]
+            private void load(AudioManager audio)
+            {
+                sample = audio.Sample.Get(@"Sliderbar/sliderbar");
+            }
+
+            private void playSample()
+            {
+                if (Clock == null)
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: `if (Clock == null || Clock.CurrentTime - lastSample <= 50) return`

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 9]
```diff
@@ -1,53 +1,163 @@
...
+                    Margin = new MarginPadding { Top = 5 },
+                    RelativeSizeAxes = Axes.X,
+                }
+            };
+        }
+
+        private class OsuSliderBar<U> : SliderBar<U> where U : struct
+        {
+            private AudioSample sample;
+            private double lastSample;
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: lastSampleTime

File: osu.Game/Overlays/Options/SliderOption.cs [hunk 10]
```diff
@@ -1,53 +1,163 @@
...
+                Height = 22;
+                Children = new Drawable[]
+                {
+                    leftBox = new Box
+                    {
+                        Height = 2,
+                        RelativeSizeAxes = Axes.None,
+                        Anchor = Anchor.CentreLeft,
+                        Origin = Anchor.CentreLeft,
+                        Colour = new Color4(255, 102, 170, 255),
```
Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: at very leave make the colour a static or member variable. it's used in 3.75 places.

Comment (peppy) [inline osu.Game/Overlays/Options/SliderOption.cs]: ColourPalette? sure, go for it. needs to be done ASAP.

Comment (SirCmpwn) [inline osu.Game/Overlays/Options/SliderOption.cs]: I do not have all of the colors, but I can work with the colors we're already using in a seperate PR

**Pull request merged**